File: backend/src/main/scala/shared/access/AccessSubject.scala
package shared.access

enum AccessSubject:
  case User(username: AccessUsername)
  case UserGroup(slug: AccessUserGroupSlug)

object AccessSubject:
  def subjectKind(value: AccessSubject): String =
    value match
      case AccessSubject.User(_) => "user"
      case AccessSubject.UserGroup(_) => "user_group"

  def subjectKey(value: AccessSubject): String =
    value match
      case AccessSubject.User(username) => username.value
      case AccessSubject.UserGroup(slug) => slug.value

File: frontend/src/shared/model/access/AccessSubject.ts
import type { AccessUserGroupSlug } from '@/shared/model/access/AccessUserGroupSlug'
import type { AccessUsername } from '@/shared/model/access/AccessUsername'

export type UserAccessSubject = {
  kind: 'user'
  username: AccessUsername
}

export type UserGroupAccessSubject = {
  kind: 'user_group'
  slug: AccessUserGroupSlug
}

export type AccessSubject = UserAccessSubject | UserGroupAccessSubject
